File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/ios/ongoing_assertion.ts
/**
 * Per-call App Attest assertion verifier used after registration.
 *
 * Registration (verifyiOSAuth) runs the full attestation chain once. Every
 * subsequent authenticated call (attestation/verify / session/token / liveness/digest) re-runs
 * a fresh assertion against the persisted credCert's public key:
 *
 *   clientDataHash = SHA-256(blob)               // blob = bytes the call signs
 *   nonce          = SHA-256(authData || clientDataHash)
 *   ECDSA_with_SHA256_verify(credCertPubKey, signature, nonce) == true
 *
 * The auth-cert ECDSA signature alone proves possession of the key but not
 * Apple-hardware provenance; this assertion binds Apple back into the loop.
 * signCount monotonicity prevents replay.
 *
 * `getExpectedIosRpIdHash` caches SHA-256(IOS_APP_ID) per process because both
 * registration and every ongoing call need it.
 */
import crypto from 'crypto';
import { parseAppAttestAssertion, parseAssertionAuthData } from './parsers';
import { getAttestationConfig } from '../config';
import type { AppAttestAssertionObject, OngoingAssertionResult } from './types';

let cachedExpectedRpIdHash: { appId: string; hash: Buffer } | null = null;

/**
 * SHA-256(IOS_APP_ID) — used by both verifyiOSAuth (registration) and
 * verifyIosOngoingAssertion (every subsequent call). Cached per process so
 * we don't rehash the same string on every request.
 */
export function getExpectedIosRpIdHash(): Buffer | null {
  const appId = getAttestationConfig().iosAppId;
  if (!appId) return null;
  if (cachedExpectedRpIdHash?.appId === appId) return cachedExpectedRpIdHash.hash;
  const hash = crypto.createHash('sha256').update(appId, 'utf8').digest();
  cachedExpectedRpIdHash = { appId, hash };
  return hash;
}

/**
 * Verify a fresh App Attest assertion produced by DCAppAttestService.generateAssertion
 * against the credCert that was persisted at registration time. Used by
 * attestation/verify / session/token / liveness/digest to re-exercise Apple hardware on every
 * authenticated call (the auth-cert ECDSA signature alone proves possession of
 * the key but not provenance; this binds Apple back into the loop).
 *
 *   clientDataHash = SHA-256(blob)            // blob = bytes the auth-cert sig signs
 *   nonce          = SHA-256(authData || clientDataHash)
 *   ECDSA_with_SHA256_verify(credCertPubKey, signature, nonce) == true
 *
 * @param credCertPem    PEM credCert from the AppAttestVerdict persisted at registration.
 * @param blob           Exact bytes the auth-cert signature signs on this call.
 * @param assertionB64   Base64 CBOR result of DCAppAttestService.generateAssertion.
 * @param expectedRpIdHash  SHA-256(IOS_APP_ID), from getExpectedIosRpIdHash().
 * @param lastSignCount  Highest signCount seen for this credCert (registration-time count
 *                       on first call). signCount must be strictly greater.
 */
export function verifyIosOngoingAssertion(
  credCertPem: string,
  blob: Buffer,
  assertionB64: string,
  expectedRpIdHash: Buffer,
  lastSignCount: number,
): OngoingAssertionResult {
  let credCertPubKey: crypto.KeyObject;
  try {
    credCertPubKey = new crypto.X509Certificate(credCertPem).publicKey;
  } catch (e) {
    return {
      ok: false,
      reason: 'CRED_CERT_PARSE_FAIL',
      message: `Failed to parse persisted credCert PEM: ${e instanceof Error ? e.message : 'unknown'}`,
    };
  }

  let parsed: AppAttestAssertionObject;
  try {
    parsed = parseAppAttestAssertion(assertionB64);
  } catch (e) {
    return {
      ok: false,
      reason: 'ASSERTION_DECODE_ERROR',
      message: `Failed to decode App Attest assertion: ${e instanceof Error ? e.message : 'unknown'}`,
    };
  }

  const authData = parsed.authenticatorData;
  const { rpIdHash, signCount } = parseAssertionAuthData(authData);

  if (!rpIdHash.equals(expectedRpIdHash)) {
    return {
      ok: false,
      reason: 'ASSERTION_RPID_MISMATCH',
      message: 'Assertion rpIdHash does not match SHA-256(IOS_APP_ID)',
      signCount,
    };
  }

  if (signCount <= lastSignCount) {
    return {
      ok: false,
      reason: 'ASSERTION_SIGNCOUNT_NOT_INCREMENTED',
      message: `Assertion signCount (${signCount}) is not greater than lastSignCount (${lastSignCount})`,
      signCount,
    };
  }

  const clientDataHash = crypto.createHash('sha256').update(blob).digest();
  const nonce = crypto
    .createHash('sha256')
    .update(authData)
    .update(clientDataHash)
    .digest();

  let signatureEncoding: 'der' | 'ieee-p1363' = 'der';
  let signatureOk = false;
  try {
    signatureOk = crypto.verify('sha256', nonce, credCertPubKey, parsed.signature);
    if (!signatureOk) {
      signatureEncoding = 'ieee-p1363';
      signatureOk = crypto.verify(
        'sha256',
        nonce,
        { key: credCertPubKey, dsaEncoding: 'ieee-p1363' },
        parsed.signature,
      );
    }
  } catch (e) {
    return {
      ok: false,
      reason: 'ASSERTION_VERIFY_EXCEPTION',
      message: `Assertion ECDSA verify threw: ${e instanceof Error ? e.message : 'unknown'}`,
      signCount,
    };
  }

  if (!signatureOk) {
    return {
      ok: false,
      reason: 'ASSERTION_SIGNATURE_INVALID',
      message: 'Assertion signature does not verify against credCert public key',
      signCount,
    };
  }

  return { ok: true, signCount, signatureEncoding };
}

File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/ios/ongoing_assertion.py
"""
Per-call App Attest assertion verifier (used after registration on every
authenticated route call).
"""

from __future__ import annotations

import hashlib
import logging
from typing import Optional

from cryptography import x509

from ..config import get_attestation_config
from ..crypto_utils import verify_signature_ec_bytes
from .parsers import parse_app_attest_assertion, parse_assertion_auth_data
from .types import OngoingAssertionResult

log = logging.getLogger("ios.ongoing_assertion")

_cached_rp_id_hash: tuple[str, bytes] | None = None


def get_expected_ios_rp_id_hash() -> Optional[bytes]:
    """SHA-256(IOS_APP_ID), cached per process. None if IOS_APP_ID is unset."""
    global _cached_rp_id_hash
    app_id = get_attestation_config().ios_app_id
    if not app_id:
        return None
    if _cached_rp_id_hash and _cached_rp_id_hash[0] == app_id:
        return _cached_rp_id_hash[1]
    h = hashlib.sha256(app_id.encode("utf-8")).digest()
    _cached_rp_id_hash = (app_id, h)
    return h


def verify_ios_ongoing_assertion(
    cred_cert_pem: str,
    blob: bytes,
    assertion_b64: str,
    expected_rp_id_hash: bytes,
    last_sign_count: int,
) -> OngoingAssertionResult:
    try:
        cred_cert_pub_key = x509.load_pem_x509_certificate(cred_cert_pem.encode("ascii")).public_key()
    except Exception as e:
        return OngoingAssertionResult(
            ok=False, reason="CRED_CERT_PARSE_FAIL",
            message=f"Failed to parse persisted credCert PEM: {e}",
        )

    try:
        parsed = parse_app_attest_assertion(assertion_b64)
    except Exception as e:
        return OngoingAssertionResult(
            ok=False, reason="ASSERTION_DECODE_ERROR",
            message=f"Failed to decode App Attest assertion: {e}",
        )

    auth_data = parsed.authenticatorData
    header = parse_assertion_auth_data(auth_data)
    rp_id_hash = header["rpIdHash"]
    sign_count = header["signCount"]

    if rp_id_hash != expected_rp_id_hash:
        return OngoingAssertionResult(
            ok=False, reason="ASSERTION_RPID_MISMATCH",
            message="Assertion rpIdHash does not match SHA-256(IOS_APP_ID)",
            signCount=sign_count,
        )
    if sign_count <= last_sign_count:
        return OngoingAssertionResult(
            ok=False, reason="ASSERTION_SIGNCOUNT_NOT_INCREMENTED",
            message=f"Assertion signCount ({sign_count}) is not greater than lastSignCount ({last_sign_count})",
            signCount=sign_count,
        )

    client_data_hash = hashlib.sha256(blob).digest()
    nonce = hashlib.sha256(auth_data + client_data_hash).digest()

    sig_ok = verify_signature_ec_bytes(nonce, parsed.signature, cred_cert_pub_key, encoding="der")
    encoding = "der"
    if not sig_ok:
        sig_ok = verify_signature_ec_bytes(nonce, parsed.signature, cred_cert_pub_key, encoding="ieee-p1363")
        encoding = "ieee-p1363"
    if not sig_ok:
        return OngoingAssertionResult(
            ok=False, reason="ASSERTION_SIGNATURE_INVALID",
            message="Assertion signature does not verify against credCert public key",
            signCount=sign_count,
        )

    return OngoingAssertionResult(ok=True, signCount=sign_count, signatureEncoding=encoding)
